File: core/src/main/java/heroes/journey/entities/Quest.java
package heroes.journey.entities;

import static heroes.journey.registries.Registries.QuestManager;

import java.util.Objects;
import java.util.function.Consumer;
import java.util.function.Predicate;

import heroes.journey.components.character.PlayerComponent;
import heroes.journey.entities.actions.inputs.ActionInput;
import heroes.journey.registries.Registrable;
import lombok.NonNull;

public class Quest extends Registrable {

    @NonNull private final Consumer<ActionInput> onComplete;
    @NonNull private final Predicate<ActionInput> isComplete;
    private final int fameReward;

    public Quest(
        String id,
        String name,
        Consumer<ActionInput> onComplete,
        Predicate<ActionInput> isComplete,
        int fameReward) {
        super(id, name);
        this.onComplete = onComplete;
        this.isComplete = isComplete;
        this.fameReward = fameReward;
    }

    public Quest(
        String id,
        String name,
        Consumer<ActionInput> onComplete,
        Predicate<ActionInput> isComplete) {
        this(id, name, onComplete, isComplete, 0);
    }

    public void onComplete(ActionInput input) {
        PlayerComponent playerComponent = PlayerComponent.get(input.getGameState().getWorld(),
            input.getEntityId());
        if (playerComponent != null) {
            playerComponent.fame(playerComponent.fame() + fameReward);
        }
        onComplete.accept(input);
    }

    public boolean isComplete(ActionInput input) {
        return isComplete.test(input);
    }

    public Quest register() {
        return QuestManager.register(this);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o)
            return true;
        if (!(o instanceof Quest quest))
            return false;
        return Objects.equals(getId(), quest.getId());
    }

    @Override
    public int hashCode() {
        return Objects.hash(getId());
    }

}

File: core/src/main/java/heroes/journey/entities/Quest.kt
package heroes.journey.entities

import heroes.journey.components.StatsComponent
import heroes.journey.entities.actions.ActionContext
import heroes.journey.entities.tagging.Attributes
import heroes.journey.modlib.Ids
import heroes.journey.modlib.attributes.Operation
import heroes.journey.modlib.misc.IQuest
import heroes.journey.modlib.registries.Registrable
import heroes.journey.mods.Registries
import heroes.journey.mods.Registries.StatManager
import java.util.*

class Quest(
    id: String,
    override val cost: Attributes = Attributes(),
    override val rewards: Attributes = Attributes(),
    override val fameReward: Int = 0
) : Registrable(id), IQuest {
    override fun register(): Quest {
        return Registries.QuestManager.register(this)
    }

    override fun canAfford(input: Any): Boolean {
        val actionContext = input as? ActionContext ?: return false
        val playerStats =
            StatsComponent.get(actionContext.gameState.getWorld(), actionContext.entityId) ?: return false
        for ((stat, requiredAmount) in cost) {
            if (playerStats[stat]!! < requiredAmount) {
                return false
            }
        }
        return true
    }

    override fun onComplete(input: Any): Boolean {
        val actionContext = input as? ActionContext ?: return false
        val playerStats = StatsComponent.get(actionContext.gameState.getWorld(), actionContext.entityId)
        if (playerStats == null) return false
        for ((stat, requiredAmount) in cost) {
            if (playerStats[stat]!! < requiredAmount) {
                return false
            }
        }
        for ((stat, amount) in cost) {
            playerStats.put(stat, -amount, Operation.ADD)
        }
        for ((stat, amount) in rewards) {
            playerStats.put(stat, amount, Operation.ADD)
        }
        if (fameReward > 0) {
            playerStats.put(StatManager[Ids.STAT_FAME], fameReward, Operation.ADD)
        }
        return true
    }

    @Suppress("PARAMETER_NAME_CHANGED_ON_OVERRIDE")
    override fun equals(o: Any?): Boolean {
        if (this === o) return true
        if (o !is Quest) return false
        return id == o.id
    }

    override fun hashCode(): Int {
        return Objects.hash(id)
    }
}
